### PR TITLE
[expo-modules][ios] Fix errors from use_frameworks in Podfile

### LINKED ISSUE #13920: Error with `pod install` using Expo 42 & `use_frameworks!`
### Summary

Enabling `use_frameworks!` with Expo 42 in a bare workflow causes `pod install` to fail because `@unimodules/core` includes statically linked binaries.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

iOS

### SDK Version (managed workflow only)

_No response_

### Environment

```
  Expo CLI 4.9.1 environment info:
    System:
      OS: macOS 11.4
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 12.22.1 - ~/.nvm/versions/node/v12.22.1/bin/node
      Yarn: 1.19.1 - /usr/local/bin/yarn
      npm: 6.14.10 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.10.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
      Android SDK:
        API Levels: 23, 25, 26, 27, 28, 29
        Build Tools: 23.0.1, 23.0.2, 25.0.2, 27.0.3, 28.0.0, 28.0.2, 28.0.3, 29.0.1, 29.0.2
        System Images: android-26 | Google APIs Intel x86 Atom
    IDEs:
      Android Studio: 4.0 AI-193.6911.18.40.6626763
      Xcode: 12.5.1/12E507 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~42.0.1 => 42.0.3
      react: 16.13.1 => 16.13.1
      react-dom: 16.13.1 => 16.13.1
      react-native: ~0.63.4 => 0.63.4
      react-native-web: ~0.13.12 => 0.13.18
    npmGlobalPackages:
      expo-cli: 4.9.1
    Expo Workflow: bare
```

### Reproducible demo or steps to reproduce from a blank project

Demo project created using the steps below:
https://github.com/carsondarling/expo42_use_frameworks_error_demo
 
### Steps to Reproduce

1. `expo init --npm --template bare-minimum`
2. Update Podfile to enable `use_frameworks!`:
```ruby
# ios/Podfile
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/react-native-unimodules/cocoapods.rb'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

platform :ios, '11.0'

target 'expo42useframeworkserrordemo' do
  use_frameworks!
  use_unimodules!
  config = use_native_modules!

  use_react_native!(:path => config["reactNativePath"])

  # Uncomment to opt-in to using Flipper
  #
  # if !ENV['CI']
  #   use_flipper!('Flipper' => '0.75.1', 'Flipper-Folly' => '2.5.3', 'Flipper-RSocket' => '1.3.1')
  #   post_install do |installer|
  #     flipper_post_install(installer)
  #   end
  # end
end
```
2. `cd ios && pod install`
3. Pod install fails with the following error: `[!] The 'Pods-expo42useframeworkserrordemo' target has transitive dependencies that include statically linked binaries: (/Users/cdarling/Workspace/expo42_use_frameworks_error_demo/node_modules/@unimodules/core/ios/UMCore.xcframework)`

```
Ignoring digest-crc-0.6.3 because its extensions are not built. Try: gem pristine digest-crc --version 0.6.3
Installing unimodules:
 expo-application@3.2.0 from ../node_modules/expo-application/ios
 expo-constants@11.0.1 from ../node_modules/expo-constants/ios
 expo-error-recovery@2.2.0 from ../node_modules/expo-error-recovery/ios
 expo-file-system@11.1.3 from ../node_modules/expo-file-system/ios
 expo-font@9.2.1 from ../node_modules/expo-font/ios
 expo-image-loader@2.2.0 from ../node_modules/expo-image-loader/ios
 expo-keep-awake@9.2.0 from ../node_modules/expo-keep-awake/ios
 expo-modules-core@0.2.0 from ../node_modules/expo-modules-core/ios
 expo-splash-screen@0.11.2 from ../node_modules/expo-splash-screen/ios
 expo-structured-headers@1.1.1 from ../node_modules/expo-structured-headers/ios
 expo-updates@0.8.3 from ../node_modules/expo-updates/ios
 expo-updates-interface@0.2.2 from ../node_modules/expo-updates-interface/ios
 unimodules-app-loader@2.2.0 from ../node_modules/unimodules-app-loader/ios
 unimodules-core@7.1.1 from ../node_modules/@unimodules/core/ios
 unimodules-react-native-adapter@6.3.5 from ../node_modules/@unimodules/react-native-adapter/ios
 unimodules-task-manager-interface@6.2.0 from ../node_modules/unimodules-task-manager-interface/ios

Auto-linking React Native modules for target `expo42useframeworkserrordemo`: RNGestureHandler, RNReanimated, and RNScreens
Analyzing dependencies
Downloading dependencies
[!] The 'Pods-expo42useframeworkserrordemo' target has transitive dependencies that include statically linked binaries: (/Users/cdarling/Workspace/expo42_use_frameworks_error_demo/node_modules/@unimodules/core/ios/UMCore.xcframework)
```

### PR DESCRIPTION
**Pull request by Kudo**
# Why

close [ENG-1772](https://linear.app/expo/issue/ENG-1772/regression-test-for-use-frameworks-in-podfile)
fixes #13920

# How

- error: `The 'App' target has transitive dependencies that include static binaries`
declare `static_framework = true` in podspec. since all expo-modules ship in either source code or static xcframework, it should make sense to add `static_framework` declaration and solve this kind of errors.

- error: generated swift runtime header not found
fix like this because in framework mode, generated headers are inside module.
```diff
+#if __has_include(<ExpoModulesCore/ExpoModulesCore-Swift.h>)
+// For cocoapods framework, the generated swift header will be inside ExpoModulesCore module
+#import <ExpoModulesCore/ExpoModulesCore-Swift.h>
+#else
 #import "ExpoModulesCore-Swift.h"
+#endif
```

- error: [gl] `<cassert>` import not found
this is supported in c++, thus rename to objc++ .mm files.

- error: [gl] `BLOCK_SAFE_RUN` inconsistent left/right operands types
this is side effect after renaming expo-gl files as .mm. fixed by explicitly cast as void.

- error: [core] undefined symbols in linking time, e.g. `EXRegisterModule`
this is side effect after renaming expo-gl files as .mm. fixed by export symbols by `extern "C"`, otherwise c++ will have different name mangling between c.

- rename `expo-image` pod to `EXImage`

- error: `Include of non-modular header inside framework module`
adding `DEFINES_MODULES` for cocoapods to create modulemap file. for external dependency like firebase, ignore the error by adding `CLANG_ALLOW_NON_MODULAR_INCLUDES_IN_FRAMEWORK_MODULES`

- error: expo-constant's app.config and expo-updates' app.manifest not found
in use_frameworks mode, the path is the same with non  framework mode. `/$UNLOCALIZED_RESOURCES_FOLDER_PATH` is unnecessary. 

- error: [barcode-scanner] ZXingObjC podspec GCC_PREPROCESSOR_DEFINITIONS does not export to EXBarCodeScanner and cause headers not found
add these defines explicitly

- [templates] add missing `react_native_post_install` from react-native 0.64 upgrade

- error: `Cycle inside FBReactNativeSpec; building could produce unreliable results. This usually can be resolved by moving the shell script phase '[CP-User] Generate Specs' so that it runs before the build phase that depends on its outputs.`
workaround by moving `FBReactNativeSpec` build phase script order.

# Test Plan

## prebuilt xcframework test

```sh
expo init -t /path/to/expo/templates/expo-template-bare-minimum sdk43
cd sdk43
# add `s.static_framework = true` for each `node_modules/expo-*/ios/*.podspec`
cd ios
# add `use_frameworks!` to Podfile
rm -rf Pods && pod install
cd ..
expo run:ios
expo run:ios --configuration Release
```

## regression for all expo-modules

```sh
expo init -t /path/to/expo/templates/expo-template-bare-minimum sdk43
cd sdk43
yarn add file:/path/to/expo/packages/expo-* # add all expo-modules one-by-one
# test non framework mode
expo run:ios
expo run:ios --configuration Release
# test framework mode
cd ios
# add `use_frameworks!` to Podfile
rm -rf Pods && pod install
cd ..
expo run:ios
expo run:ios --configuration Release
```

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by Kudo** [comment]
ci looks fine except dev-client e2e which is also broken on master. will write a script to update changelog for all modules tomorrow.

**Comment by Kudo** [comment]
another known issue is that hermes with use_frameworks! breaking in react-native 0.64 😖

File: packages/expo-image/ios/EXImage.podspec [hunk 1]
```diff
@@ -20,7 +20,7 @@ if using_custom_sd_web_image_webp_coder_version
 end
 
 Pod::Spec.new do |s|
-  s.name           = 'expo-image'
+  s.name           = 'EXImage'
```
Comment (tsapeta) [inline packages/expo-image/ios/EXImage.podspec:23]: We will do another rename to `ExpoImage` when I rewrite it to Swift 😄 For now I'm fine with `EXImage`.

File: packages/expo-image/ios/EXImage.podspec [hunk 2]
```diff
@@ -1,6 +1,6 @@
 require 'json'
 
-package = JSON.parse(File.read(File.join(__dir__, 'package.json')))
+package = JSON.parse(File.read(File.join(__dir__, '..', 'package.json')))
```
Comment (tsapeta) [inline packages/expo-image/ios/EXImage.podspec:3]: Are you sure `expo-image` will be autolinked by RN when you move the podspec? As far as I remember, the CLI looks for them only in the root dir if not specified in `react-native.config.js`.

**Review by tsapeta — COMMENTED** [review]
Wow, good job! 👏 Left a few comments tho 👇

File: packages/expo-dev-launcher/ios/EXDevLauncher.m [hunk 3]
```diff
@@ -3,7 +3,12 @@
 #import "EXDevLauncher.h"
 #import "EXDevLauncherController.h"
 
+#if __has_include(<EXDevLauncher/EXDevLauncher-Swift.h>)
+// For cocoapods framework, the generated swift header will be inside EXDevLauncher module
+#import <EXDevLauncher/EXDevLauncher-Swift.h>
+#else
 #import <EXDevLauncher-Swift.h>
+#endif
```
[the remark this answers is not in the document]
Comment (Kudo) [reply packages/expo-dev-launcher/ios/EXDevLauncher.m:11]: we can try it later, but i am afraid if that is truly feasible. this macro will be an external dependency to other package (e.g. in `expo-modules-core`). at least for prebuilding xcframework, so far i didn't find a way to have external dependency in headers 😅

File: packages/expo-image/ios/EXImage.podspec (hunk 1 above)
Comment (Kudo) [reply packages/expo-image/ios/EXImage.podspec:23]: it sounds good. there are too many EX prefixed stuffs, please help to do the rename together during the swift rewrite.

File: packages/expo-image/ios/EXImage.podspec (hunk 2 above)
Comment (Kudo) [reply packages/expo-image/ios/EXImage.podspec:3]: good catch 👍  i didn't notice expo-image using cli linking. i'll move the podspec back.

**Comment by Kudo** [comment]
leaving note for bare-expo with use_frameworks! still breaking with these reasons:

1. does not works with [`generate_multiple_pod_projects`](https://github.com/expo/expo/blob/f31f564bde2e15fa99d8fb3303ddf3fd6f2c0a3e/apps/bare-expo/ios/Podfile#L1-L3)

2. some other vendor modules should revise `spec.dependency 'React'` to `spec.dependency 'React-Core'` as https://github.com/facebook/react-native/issues/29633#issuecomment-694187116 mentioned.
- react-native-svg: should upgrade to use @next version
- [react-native-shared-element](https://github.com/IjzerenHein/react-native-shared-element/blob/edd7c5b32eac8197e736ae91b360d98de1edabaa/RNSharedElement.podspec#L22): i can send a pr next week
- react-native-view-shot: upstream has the fix already but not yet have a new release on npm
- react-native-appearance: upstream has the fix already but not yet have a new release on npm

**Review by tsapeta — APPROVED** [review]

**Pull request merged**